File: ex4nicegui/gsap/timeline.js
import { gsap } from "gsap";
import { convertDynamicProperties } from "../../static/utils/dynamic_properties.js";


export default {
  template: `<template></template>`,
  data() {

    const tl = gsap.timeline(this.defaults)
    this.tl = tl

    return {
    }

  },
  mounted() {
    document.addEventListener('DOMContentLoaded', () => {
      /**
       * @type any[]
       */
      const tasks = this.tasks
      tasks.forEach(t => {
        this[t.method](t.targets, t.vars, t.position)
      })

      this.scriptTasks.forEach(script => {
        const fn = new Function('return ' + script)()
        fn(this.tl, gsap)
      })

    })
  },

  methods: {
    from(targets, vars, position) {
      convertDynamicProperties(vars, false)
      this.tl.from(targets, vars, position)
    },
    to(targets, vars, position) {
      convertDynamicProperties(vars, false)
      this.tl.to(targets, vars, position)
    },
    runScript(script) {
      const fn = new Function('return ' + script)()
      fn(this.tl, gsap)
    },
    callTimeline(name, ...args) {
      this.tl[name](...args)
    }
  },
  props: {
    defaults: Object,
    tasks: Array,
    scriptTasks: Array,
  },
};

File: ex4nicegui/gsap/timeline.py
import os
from pathlib import Path
from typing import Dict, Optional, Union
from nicegui.element import Element
from nicegui import context as ng_context


class Timeline(
    Element,
    component="timeline.js",
    exposed_libraries=["../libs/gsap/gsap.mjs"],
    extra_libraries=["../libs/gsap/*.js", "../libs/gsap/utils/*.js"],
):
    def __init__(
        self,
        defaults: Optional[Dict] = None,
    ) -> None:
        super().__init__()
        self._props["defaults"] = defaults or {}
        self._props["tasks"] = []
        self._props["scriptTasks"] = []

    def from_(self, targets: str, vars: Dict, position: Optional[str] = None):
        self.__try_run_task("from", targets, vars, position)
        return self

    def to(self, targets: str, vars: Dict, position: Optional[str] = None):
        self.__try_run_task("to", targets, vars, position)
        return self

    def run_script(self, script: Union[str, Path]):
        if isinstance(script, Path):
            script = script.read_text("utf8")

        script = os.linesep.join([s for s in script.splitlines() if s])
        self.__try_run_script_task(script)
        return self

    def __try_run_script_task(self, script: str):
        def fn():
            self.run_method("runScript", script)

        if ng_context.get_client().has_socket_connection:
            fn()
        else:
            tasks = self._props["scriptTasks"]
            tasks.append(script)

    def __try_run_task(
        self, name: str, targets: str, vars: Dict, position: Optional[str] = None
    ):
        def fn():
            self.run_method(name, targets, vars)

        if ng_context.get_client().has_socket_connection:
            fn()
        else:
            tasks = self._props["tasks"]
            tasks.append(
                {"method": name, "targets": targets, "vars": vars, "position": position}
            )

    def pause(self):
        self.run_method("callTimeline", "pause")

    def play(self):
        self.run_method("callTimeline", "play")

    def resume(
        self,
    ):
        self.run_method("callTimeline", "resume")

    def seek(self, position: Optional[str] = None, suppressEvents=True):
        self.run_method("callTimeline", "seek", position, suppressEvents)

    def reverse(self):
        self.run_method("callTimeline", "reverse")
